### ISSUE TITLE
Changes in models between versions

### ISSUE THREAD

**Issue by eadstry** [ydchen0806/CLC#1]
I noticed that you recently resubmitted the code for CLC_run.py, which is quite different from the original implementation, such as:
In the initial version，
The reference feature needs to be fused with y

>         
        y = self.g_a(x)
        y_shape = y.shape[2:]
        if x_refs is not None:
            target_size = x.size()[2:]  
            resized_ref_x_list = [
            F.interpolate(ref_x, size=target_size, mode='bilinear', align_corners=False) 
            for ref_x in x_refs
            ]
            y_refs = [self.g_a(x_ref) for x_ref in resized_ref_x_list]
            # y_refs = torch.stack(y_refs, dim=1)
            y_refs_aligned = self.clm(y, y_refs)
            # y_refs_aligned = [self.clm(y, y_ref) for y_ref in y_refs]
            y = self.cls_compress(y, y_refs_aligned)
And the reference features do not directly participate in the prediction of distribution parameters.

> 
            support_slices = (y_hat_slices if self.max_support_slices < 0 else y_hat_slices[:self.max_support_slices])
            mean_support = torch.cat([latent_means] + support_slices, dim=1)
            mean_support = self.atten_mean[slice_index](mean_support)
            mu = self.cc_mean_transforms[slice_index](mean_support)

Then, in the latest version, the reference feature is no longer fused with y, but participates in the prediction of distribution parameters.

> 
       ref_features = self.extract_ref_features(ref_frames)
        
        # Main encoding process
        y = self.g_a(x)
        y_shape = y.shape[2:]
        z = self.h_a(y)
        _, z_likelihoods = self.entropy_bottleneck(z)

        z_offset = self.entropy_bottleneck._get_medians()
        z_tmp = z - z_offset
        z_hat = ste_round(z_tmp) + z_offset

>            
            if ref_features is not None and self.use_ref:
                # Combine reference features with main features
                mean_support_with_ref = torch.cat([mean_support, ref_features], dim=1)
                scale_support_with_ref = torch.cat([scale_support, ref_features], dim=1)
                
                # Use reference-enhanced transforms
                mu = self.ref_cc_mean_transforms[slice_index](mean_support_with_ref)
                scale = self.ref_cc_scale_transforms[slice_index](scale_support_with_ref)
            else:
                # Use original transforms (backward compatibility)
                mu = self.cc_mean_transforms[slice_index](mean_support)
                scale = self.cc_scale_transforms[slice_index](scale_support)
I think these two implementations are very different. Can you tell me what this change means?

**Comment by ydchen0806** [comment]
Hi, thank you for your interest in our project. Apologies for the previous version I uploaded—it was based on the latest method we're working on for reference images (expanded from the AAAI version to the journal). This method aims to make the reference image more similar to the currently encoded image and better aligned at the structural level, allowing us to directly fuse them at the feature level. The officially released version (matching the weights in HF) uses a simple transformation for the reference image, mainly applying it to the entropy model. This allows for greater tolerance of reference images, even if they aren't perfectly aligned structurally, and still provides some gains (though the performance might be slightly lower).

**Comment by eadstry** [comment]
The officially released version does not seem to use the CLM modules mentioned in the paper, which is what confuses me.  
In addition, it seems that the sample is resized in dataset's `__getitem__ ` but is only used for `extract_feature`, while the returned sample uses `normalize_to_tensor`, which causes an error during training.

**Comment by ydchen0806** [comment]
Hello,

Thank you for your feedback. In the code, the `class CLM(nn.Module)` defines the alignment and fusion modules mentioned in the paper. Specifically, the `alignment` module performs feature alignment using attention, and the `fusion` module is responsible for feature fusion, as explained in the comments in the code.

Regarding the issue with resizing in the `__getitem__` method, `extract_feature` is used to extract image features, which are then used as queries to match against the reference image dictionary. The `normalize_to_tensor` function normalizes the reference image values to the range of 0 to 1 and sets the channels-first format, which is required for training. This should help avoid errors during training.

Regarding resizing the sample, I believe you're referring to the pooling operation we use during training to make the reference image size match the input `x` size. Of course, we recommend using reference images and `x` with the same size for best results.

If you need further clarification, we can share the code and weights for both the conference and journal papers (currently under submission) for your reference.
